File: prettier-plugin-csharpier/src/Printer/Types/SwitchSection.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { printBlock } from "./Block";

export interface SwitchSectionNode extends SyntaxTreeNode<"SwitchSection"> {
    labels: SyntaxTreeNode[];
    statements: SyntaxTreeNode[];
}

export const printSwitchSection: PrintMethod<SwitchSectionNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [join(hardline, path.map(print, "labels"))];
    if (node.statements.length === 1 && node.statements[0].nodeType === "Block") {
        parts.push(path.call(o => printBlock(o, options, print), "statements", 0));
    } else {
        parts.push(indent(concat([hardline, concat(path.map(print, "statements"))])));
    }
    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/SwitchSection.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class SwitchSection
{
    public static Doc Print(SwitchSectionSyntax node, PrintingContext context)
    {
        var docs = new ValueListBuilder<Doc>([null, null]);
        docs.Append(Doc.Join(Doc.HardLine, node.Labels.Select(o => Node.Print(o, context))));
        if (node.Statements is [BlockSyntax blockSyntax])
        {
            docs.Append(Block.Print(blockSyntax, context));
        }
        else
        {
            docs.Append(
                Doc.Indent(
                    node.Statements.First() is BlockSyntax ? Doc.Null : Doc.HardLine,
                    Doc.Join(
                        Doc.HardLine,
                        node.Statements.Select(o => Node.Print(o, context)).ToArray()
                    )
                )
            );
        }
        return Doc.Concat(ref docs);
    }
}
